src/apps/tap/tap.lua

```diff
@@ -0,0 +1,99 @@
+module(..., package.seeall)
+
+local S = require("syscall")
+local link = require("core.link")
+local packet = require("core.packet")
+local ffi = require("ffi")
+local C = ffi.C
+local mac = require("lib.macaddress")
+local const = require "syscall.linux.constants"
+
+local t = S.types.t
+
+Tap = { }
+
+--- Where should this go?
+local SIOCSIFHWADDR = 0x8924
+
+function Tap:new (cfg)
+   if type(cfg) == "string" then
+      cfg = { name = cfg }
+   end
+   -- Are assertions the way to signal brokenness upwards?
+   assert(cfg['name'], "missing tap interface name")
+   
+   local sock, err = S.open("/dev/net/tun", "rdwr, nonblock");
+   assert(sock, "Error opening /dev/net/tun: " .. tostring(err))
+   local ifr = t.ifreq()
+   ifr.flags = "tap, no_pi"
+   ifr.name = cfg['name']
+   local ok, err = sock:ioctl("TUNSETIFF", ifr)
+   if not ok then
+      S.close(sock)
+      assert(false, "Error opening /dev/net/tun: " .. tostring(err))
+   end
+   
+   local obj = setmetatable({sock = sock}, {__index = Tap})
+   
+   if cfg['mac'] then
+      obj:mac(cfg['mac'])
+   end
+   return obj
+end
+
+function Tap:mac (macaddr)
+   local macaddr,err = mac:new(macaddr)
+   assert(macaddr, tostring(err))
+
+   local ifr = t.ifreq()
+   ifr.ifr_ifru.ifru_hwaddr.sa_family = 1
+   local m = ffi.cast('uint8_t*', ifr.ifr_ifru.ifru_hwaddr.sa_data)
+   ffi.copy(m, macaddr.bytes, mac.ETHER_ADDR_LEN)
+   local ok, err = self.sock:ioctl(SIOCSIFHWADDR, ifr)
+   assert(ok, tostring("Failed to set mac address: " .. tostring(err)))
+end
+
+function Tap:pull ()
+   local l = self.output.output
+   if l == nil then return end
+   while not link.full(l) do
+      local p = packet.allocate()
+      local len, err = S.read(self.sock, p.data, C.PACKET_PAYLOAD_SIZE)
+      if not len and err.errno == const.E.AGAIN then 
+         packet.free(p)
+         return
+      end
+      if not len then
+         -- How should this be bubbled up?
+         assert(0, "something broke")
+      end
```

Comment:
Good question! Handling unexpected I/O errors is not something we handle very sophisticatedly today.

Raising an error (like this assert) is probably a reasonable strategy. Really we need to restore the engine code that catches these and restarts individual apps. Restarting the app does seem like the most reasonable response to an unrecognized error (e.g. maybe somebody manually deleted the tap device with `tunctl` and it needs to be recreated). This would be the normal [crash-only software](https://en.wikipedia.org/wiki/Crash-only_software) or [let it crash](http://stackoverflow.com/questions/4393197/erlangs-let-it-crash-philosophy-applicable-elsewhere) design applied at the app level.

For non-fatal errors like the `EAGAIN` we should also bump `core.counter` counters. Today we are pretty slack with counters and that is a pity because they are crucial for understanding what is going on with a running system. I think that each app should have a well-defined set of counters (`tx_dropped`, etc) as part of its interface and documentation on equal footing to its configuration items. Maybe the Tap app could be an early adopter :-) up to you.